cpp/turbodbc_arrow/Library/src/arrow_result_set.cpp

```diff
@@ -0,0 +1,258 @@
+#include <turbodbc_arrow/arrow_result_set.h>
+
+// Somewhere a macro defines BOOL as a constant. This is in conflict with array/type.h
+#undef BOOL
+#include <arrow/array.h>
+#include <arrow/builder.h>
+#include <arrow/column.h>
+#include <arrow/schema.h>
+#include <arrow/table.h>
+#include <arrow/type.h>
+#include <arrow/memory_pool.h>
+
+#include <pyarrow/table_api.h>
+
+#include <sql.h>
+
+#include <vector>
+
+using arrow::default_memory_pool;
+using arrow::ArrayBuilder;
+using arrow::BooleanBuilder;
+using arrow::DateBuilder;
+using arrow::DoubleBuilder;
+using arrow::Int64Builder;
+using arrow::Status;
+using arrow::StringBuilder;
+using arrow::TimeUnit;
+using arrow::TimestampBuilder;
+
+namespace turbodbc_arrow {
+
+
+namespace {
+
+	std::unique_ptr<ArrayBuilder> make_array_builder(turbodbc::type_code type)
+	{
+		switch (type) {
+			case turbodbc::type_code::floating_point:
+				return std::unique_ptr<ArrayBuilder>(new DoubleBuilder(default_memory_pool(), arrow::float64()));
+			case turbodbc::type_code::integer:
+				return std::unique_ptr<ArrayBuilder>(new Int64Builder(default_memory_pool(), arrow::int64()));
+			case turbodbc::type_code::boolean:
+				return std::unique_ptr<ArrayBuilder>(new BooleanBuilder(default_memory_pool(), std::make_shared<arrow::BooleanType>()));
+			case turbodbc::type_code::timestamp:
+				return std::unique_ptr<TimestampBuilder>(new TimestampBuilder(default_memory_pool(), arrow::timestamp(TimeUnit::MICRO)));
+			case turbodbc::type_code::date:
+        return std::unique_ptr<DateBuilder>(new DateBuilder(default_memory_pool(), arrow::date()));
+			default:
+				return std::unique_ptr<StringBuilder>(new StringBuilder(default_memory_pool()));
+		}
+	}
+
+	int64_t timestamp_to_microseconds(char const * data_pointer)
+	{
+		auto & sql_ts = *reinterpret_cast<SQL_TIMESTAMP_STRUCT const *>(data_pointer);
+		long const microseconds = sql_ts.fraction / 1000;
+    struct tm datetime = {0};
+    datetime.tm_year = sql_ts.year - 1900;
+    datetime.tm_mon = sql_ts.month - 1;
+    datetime.tm_mday = sql_ts.day;
+    datetime.tm_hour = sql_ts.hour;
+    datetime.tm_min = sql_ts.minute;
+    datetime.tm_sec = sql_ts.second;
+    struct tm epoch = {0};
+    epoch.tm_year = 70;
+    epoch.tm_mday = 1;
+
+    // Microseconds since the epoch
+    return lrint(difftime(mktime(&datetime), mktime(&epoch))) * 1000000 + microseconds;
+	}
+
+	int64_t date_to_days(char const * data_pointer)
+	{
+		auto & sql_date = *reinterpret_cast<SQL_DATE_STRUCT const *>(data_pointer);
+
+    struct tm date = {0};
+    date.tm_year = sql_date.year - 1900;
+    date.tm_mon = sql_date.month - 1;
+    date.tm_mday = sql_date.day;
```

Comment:
Should this not also have -1? since the definition for std::tm is "since 1900 01 01"?